### ISSUE TITLE
WebSharper 3.0.15.92-alpha crashes on Windows Phone 8.1

### ISSUE THREAD

**Issue by ca-ta** [intellifactory/websharper#330]
The previous version, (WebSharper 3.0.12.88-alpha), works just fine.

There error (as displayed in Visual Studio) is: 0x800a139e - JavaScript runtime error: InvalidStateError, the line of code is "xhr.withCredentials=true;" in the snippet below (part of the Websharper runtime)

     ajax:function($async,$url,$headers,$data,$ok,$err)
     {
      var $0=this,$this=this;
      var xhr=new Global.XMLHttpRequest();
      xhr.withCredentials=true;
      xhr.open("POST",$url,$async);
      for(var h in $headers){
       xhr.setRequestHeader(h,$headers[h]);
      }

This is all I got at the moment - do you guys have any idea what might be causing this? It appears related to the RPC implementation, but I do not know enough about Websharper to make any intelligent guess as to why it broke.

Let me know if I can be of any help - Websharper not working on Windows Phone is a deal breaker for us.

**Comment by Tarmil** [comment]
According to [this](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest), it seems that `withCredentials` should not be used for synchronous requests. Are you using a synchronous RPC (ie. an RPC function whose return type is neither `unit` nor `Async<_>`)? If so, I highly recommend that you use an asynchronous request instead: with synchronous requests, the user interface can get frozen until the response arrives, which greatly worsens the user experience -- especially on mobile.

In any case, if this is indeed the issue, then I will add a test in this function so that `withCredentials` is only set if the request is asynchronous.

**Comment by ca-ta** [comment]
Hi Tarmil, thank you for getting back to me.

Well, all our RPC requests *are* asynchronous - we're in agreement that synchronous calls over the are generally a very bad idea :). Thank you for the link to the spec - it helped tremendously.

It seems that WP8 browser is very strict about that particular piece of spec, and throws an exception if one attempts to set ```withCredentials``` on a request that has not been opened in asynchronous mode. So, in fact, no RPC calls get made.

Long story short:
1. Yes, it's worth checking whether the request is async and only set ```withCredentials```  if it is.
2. Only set ```withCredentials```  on an opened async ```XMLHttpRequest```.

I've submitted a pull request - see #331

**Comment by Tarmil** [comment]
Thanks for looking more into this and for the PR!
